pkg/tagset/composite_tags.go

```diff
@@ -0,0 +1,121 @@
+}
+
+// NewCompositeTags creates a new CompositeTags
+func NewCompositeTags(tags1 []string, tags2 []string) *CompositeTags {
+	return &CompositeTags{
+		tags1: tags1,
+		tags2: tags2,
+	}
+}
+
+// CompositeTagsFromSlice creates a new CompositeTags from a slice
+func CompositeTagsFromSlice(tags []string) *CompositeTags {
+	return NewCompositeTags(tags, nil)
+}
+
+// CombineCompositeTagsAndSlice creates a new CompositeTags from an existing CompositeTags and string slice
+func CombineCompositeTagsAndSlice(compositeTags *CompositeTags, tags []string) *CompositeTags {
+	if compositeTags.tags2 == nil {
+		return NewCompositeTags(compositeTags.tags1, tags)
+	}
+	// Copy tags in case `CombineCompositeTagsAndSlice` is called twice with the same first argument.
+	// For example see TestCompositeTagsCombineCompositeTagsAndSlice.
+	newTags := append(append([]string{}, compositeTags.tags2...), tags...)
+	return NewCompositeTags(compositeTags.tags1, newTags)
+}
+
+// ForEach applies `callback` to each tag
+func (t *CompositeTags) ForEach(callback func(tag string)) {
+	for _, t := range t.tags1 {
+		callback(t)
+	}
+	for _, t := range t.tags2 {
+		callback(t)
+	}
+}
+
+// ForEachErr applies `callback` to each tag while `callback`` returns nil.
+// The first error is returned.
+func (t *CompositeTags) ForEachErr(callback func(tag string) error) error {
+	for _, t := range t.tags1 {
+		if err := callback(t); err != nil {
+			return err
+		}
+	}
+	for _, t := range t.tags2 {
+		if err := callback(t); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Find returns whether `callback` returns true for a tag
+func (t *CompositeTags) Find(callback func(tag string) bool) bool {
+	for _, t := range t.tags1 {
+		if callback(t) {
+			return true
+		}
+	}
+	for _, t := range t.tags2 {
+		if callback(t) {
+			return true
+		}
+	}
+
+	return false
+}
+
+// Len returns the length of the tags
+func (t *CompositeTags) Len() int {
+	return len(t.tags1) + len(t.tags2)
+}
+
+// Join performs strings.Join on tags
+func (t *CompositeTags) Join(separator string) string {
+	if len(t.tags2) == 0 {
+		return strings.Join(t.tags1, separator)
+	}
+	if len(t.tags1) == 0 {
+		return strings.Join(t.tags2, separator)
+	}
+	return strings.Join(t.tags1, separator) + separator + strings.Join(t.tags2, separator)
+}
+
+// MarshalJSON serialization a Payload to JSON
+func (t *CompositeTags) MarshalJSON() ([]byte, error) {
+	tags := append([]string{}, t.tags1...)
+	return json.Marshal(append(tags, t.tags2...))
+}
+
+// UnmarshalJSON implements the json.Unmarshaler interface.
+func (t *CompositeTags) UnmarshalJSON(b []byte) error {
+	t.tags2 = nil
+	return json.Unmarshal(b, &t.tags1)
+}
+
+// UnsafeToReadOnlySliceString creates a new slice containing all tags.
+// The caller of this method must ensure that the slice is never mutate.
+// Should be used only for performance reasons.
+func (t *CompositeTags) UnsafeToReadOnlySliceString() []string {
+	return append(t.tags1, t.tags2...)
```

Comment:
This should probably copy `t.tags1`, as that slice and its backing array were taken from the caller of one of the constructors, and other slices may share that backing array.  The `append` may modify elements of the backing array if it does not reallocate.

Alternately, this caveat could be mentioned in the doc comments.

Reply:
I have updated the implementation to
```go
if len(t.tags2) == 0 {
	return t.tags1
}
return append(append([]string{}, t.tags1...), t.tags2...)
```
In the context where `UnsafeToReadOnlySliceString` is called `len(t.tags2) == 0`.